config/kernel-timer.m4

```diff
@@ -1,32 +1,71 @@
 dnl #
-dnl # 4.15 API change
-dnl # https://lkml.org/lkml/2017/11/25/90
-dnl # Check if timer_list.func get passed a timer_list or an unsigned long
-dnl # (older kernels).  Also sanity check the from_timer() and timer_setup()
-dnl # macros are available as well, since they will be used in the same newer
-dnl # kernels that support the new timer_list.func signature.
+dnl # 4.14-rc3 API change
+dnl # https://lwn.net/Articles/735887/
+dnl # Check for the new timer_setup() API, which can be available with either
+dnl # the old or the new callback signature, also check for the existance of
+dnl # flags in struct timer_list, they are not present on much older kernels.
 dnl #
-AC_DEFUN([ZFS_AC_KERNEL_TIMER_FUNCTION_TIMER_LIST], [
-	AC_MSG_CHECKING([whether timer_list.function gets a timer_list])
+AC_DEFUN([ZFS_AC_KERNEL_TIMER_SETUP], [
+	AC_MSG_CHECKING([whether timer_setup() is available])
 	tmp_flags="$EXTRA_KCFLAGS"
 	EXTRA_KCFLAGS="-Werror"
+
+	have_timer_setup=
 	ZFS_LINUX_TRY_COMPILE([
 		#include <linux/timer.h>
-		void task_expire(struct timer_list *tl) {}
+
+		struct my_task_timer {
+			struct timer_list timer;
+			int data;
+		};
+
+		void task_expire(struct timer_list *tl)
+		{
+			struct my_task_timer *task_timer = from_timer(task_timer, tl, timer);
+			task_timer->data = 42;
+		}
+	],[
+		struct my_task_timer task_timer;
+		timer_setup(&task_timer.timer, task_expire, 0);
+	],[
+		AC_MSG_RESULT(yes)
+		AC_DEFINE(HAVE_KERNEL_TIMER_SETUP, 1,
+		    [timer_setup() is available])
+		have_timer_setup=1
 	],[
-		#ifndef from_timer
-		#error "No from_timer() macro"
-		#endif
+		AC_MSG_RESULT(no)
+	])
+
+	AC_MSG_CHECKING([whether timer function expects timer_list])
 
-		struct timer_list timer;
-		timer.function = task_expire;
-		timer_setup(&timer, NULL, 0);
+	ZFS_LINUX_TRY_COMPILE([
+		#include <linux/timer.h>
+		void task_expire(struct timer_list *tl) {}
+	],[
+		struct timer_list tl;
+		tl.function = task_expire;
 	],[
 		AC_MSG_RESULT(yes)
 		AC_DEFINE(HAVE_KERNEL_TIMER_FUNCTION_TIMER_LIST, 1,
 		    [timer_list.function gets a timer_list])
 	],[
 		AC_MSG_RESULT(no)
 	])
+
+	AC_MSG_CHECKING([whether struct timer_list has flags])
+
+	ZFS_LINUX_TRY_COMPILE([
+		#include <linux/timer.h>
+	],[
+		struct timer_list tl;
+		tl.flags = 2;
+	],[
+		AC_MSG_RESULT(yes)
+		AC_DEFINE(HAVE_KERNEL_TIMER_LIST_FLAGS, 1,
+		    [struct timer_list has a flags member])
+	],[
+		AC_MSG_RESULT(no)
+	])
+
```

Comment:
This change is from a commit in 4.2 kernel.
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=0eeda71bc30d74f66f8231f45621d5ace3419186